modules/apps/frontend-js/frontend-js-top-head-extender/src/main/java/com/liferay/frontend/js/top/head/extender/internal/servlet/taglib/TopHeadDynamicInclude.java

```diff
@@ -160,63 +212,79 @@ private void _addPortalBundles(List<String> urls, String propsKey) {
 		}
 	}
 
-	private synchronized void _rebuild() {
-		if ((_bundleContext == null) || (_portal == null) ||
-			(_portalWebResources == null)) {
+	private ResourceURLsBag _getResourceURLsBag() {
+		ResourceURLsBag resourceURLsBag = _resourceURLsBag;
 
-			return;
+		if (resourceURLsBag != null) {
+			return resourceURLsBag;
 		}
 
-		_allJsResourceURLs.clear();
+		synchronized (_topHeadResourcesServiceReferences) {
+			if (_resourceURLsBag != null) {
+				return _resourceURLsBag;
+			}
 
-		_addPortalBundles(
-			_allJsResourceURLs, PropsKeys.JAVASCRIPT_EVERYTHING_FILES);
+			_resourceURLsBag = _rebuild();
 
-		_jsResourceURLs.clear();
+			return _resourceURLsBag;
+		}
+	}
 
-		_addPortalBundles(_jsResourceURLs, PropsKeys.JAVASCRIPT_BAREBONE_FILES);
+	private ResourceURLsBag _rebuild() {
+		ResourceURLsBag resourceURLsBag = new ResourceURLsBag();
 
-		synchronized (_topHeadResourcesServiceReferences) {
-			for (ServiceReference<TopHeadResources>
-					topHeadResourcesServiceReference :
-						_topHeadResourcesServiceReferences) {
+		resourceURLsBag._allJsResourceURLs.clear();
```

Comment:
@slnn This is just created, why clear the list?

Reply:
The class did it https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-top-head-extender/src/main/java/com/liferay/frontend/js/top/head/extender/internal/servlet/taglib/TopHeadDynamicInclude.java#L170, I just keep same logic as before.

Reply:
@slnn But now you re-create a new `Bag`, so the list in the new `Bag` is already empty.